lesson-003-web-server-graceful-shutdown/README.md

```diff
@@ -0,0 +1,104 @@
+# Lesson 3: Web server graceful shutdown
+
+Taking as a foundation the webserver created in [lesson 000](../lesson-000-web-server/), we will be adding `graceful shutdown` functionality to it.
+
+In order to achieve this, we will be using `channels` and `goroutines`
+
+## Knowledge bits
+
+
+### Goroutines
+
+A `goroutine` is a lightweight thread of execution managed by the Go runtime; it is the way we have to run a piece of code concurrently with the original calling code, as we can see explained in the [oficial documentation.](https://tour.golang.org/concurrency/1)
+
+A small example could be:
+
+```
+func say(s string) {
+	for i := 0; i < 5; i++ {
+		time.Sleep(100 * time.Millisecond)
+		fmt.Println(s)
+	}
+}
+
+func main() {
+	go say("k8s")
+	say("contribute")
+}
+```
+
+Calling `go printMyString()` is where we start a new `goroutine`, calling our already defined `printMyString` function, while the `main` function runs in its own `goroutine` (called the _main goroutine_)
+
+
+### Channels
+
+Channels are a concurrency synchronization technique we can use in Go.
+
+We can create a channel using the keyword `chan`, and we can transport data of only one data type.
+
+From the [official documention](https://tour.golang.org/concurrency/2) we can see how data can be sent and received from the channel.
+
+```
+ch := make(chan int)
+
+ch <- v    // Send v to channel ch.
+v := <-ch  // Receive from ch, and
+           // assign value to v.
+```
+
+**_Unbuffered_ vs _buffered_ channels**
+
+Channels are created **blocking by default** (or _unbuffered_)
+
+Unbuffered channels are those which both goroutines (one sending data to the buffer, another waiting to receive the resource) require to be ready to exchange some data. If a goroutine is trying to send a resource to an unbuffered channel and there is no goroutine waiting to receive that resource, the channel will lock the sender and make it wait.
```

Comment:
```suggestion
If we try to send a resource to an unbuffered channel, the channel will lock (preventing us from sending anything else into the channel) until someone else reads from it.
This essentially means that an unbuffered channel is restricted to never having more than 1 element inside of it.
```

Reply:
Small rewrite here.
Suggesting to remove go routines from the wording since normal functions can also send and receive from channels.